### PR TITLE
EZP-24278: Add new stack kernel around REST API v1

### PR DESCRIPTION
**Pull request by bdunogier**
> Fixes http://jira.ez.no/browse/EZP-24278
> Requires ezsystems/LegacyBridge#25

This PR adds the required elements (custom kernel, mostly) to run the V1 REST API through the new stack.

index_rest.php is still used, but it uses the custom rest kernel instead.

### CODE REVIEW

File: kernel/private/rest/classes/ezpkernelrest.php [hunk 1]
```diff
@@ -0,0 +1,294 @@
+<?php
+/**
+ * File containing the ezpKernelTreeMenu class.
```
Comment (crevillo) [inline kernel/private/rest/classes/ezpkernelrest.php]: no ezpKernelTreeMenu here :)

File: index_rest.php [hunk 2]
```diff
@@ -9,70 +9,12 @@
  */
 
 require __DIR__ . '/autoload.php';
-require __DIR__ . '/kernel/private/rest/classes/lazy.php';
+// ??? require __DIR__ . '/kernel/private/rest/classes/lazy.php';
```
Comment (bdunogier) [inline index_rest.php]: This needs to be removed (done later in the Kernel).

File: kernel/private/rest/classes/ezpkernelrest.php [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+     * @var array
+     */
+    protected $access;
+
+    /**
+     * @var eZURI
+     */
+    protected $uri;
+
+    private $responseWriterClass;
```
Comment (bdunogier) [inline kernel/private/rest/classes/ezpkernelrest.php]: phpdoc

File: kernel/private/rest/classes/ezpkernelrest.php [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+    private static $response;
```
Comment (bdunogier) [inline kernel/private/rest/classes/ezpkernelrest.php]: phpdoc

File: kernel/private/rest/classes/mvc_configuration.php [hunk 5]
```diff
@@ -21,9 +21,12 @@ class ezpMvcConfiguration implements ezcMvcDispatcherConfiguration
      */
     public $apiPrefix;
 
-    public function __construct()
+    private $responseWriterClass;
```
Comment (bdunogier) [inline kernel/private/rest/classes/mvc_configuration.php]: phpdoc

Comment (bdunogier) [inline kernel/private/rest/classes/mvc_configuration.php]: phpdoc

File: kernel/private/rest/classes/ezpkernelrest.php [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+
+    public function __construct( array $settings = array(), $responseWriterClass )
```
Comment (andrerom) [inline kernel/private/rest/classes/ezpkernelrest.php]: phpdoc :)

Comment (bdunogier) [inline kernel/private/rest/classes/ezpkernelrest.php]: Actually, I don't really bother doing so for the ctor when there's nothing special to say...

File: kernel/private/rest/classes/ezpkernelrest.php [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    }
+
+    /**
+     * Not supported by ezpKernelTreeMenu
+     *
+     * @throws \RuntimeException
+     */
+    public function runCallback( \Closure $callback, $postReinitialize = true )
+    {
+        throw new \RuntimeException( 'runCallback() method is not supported by ezpKernelTreeMenu' );
```
Comment (lolautruche) [inline kernel/private/rest/classes/ezpkernelrest.php]: no treemenu here

File: kernel/private/rest/classes/ezpkernelrest.php [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+
+    /**
+     * Checks whether the kernel handler has the Symfony service container
+     * container or not.
+     *
+     * @return bool
+     */
+    public function hasServiceContainer()
+    {
+        return false;
```
Comment (lolautruche) [inline kernel/private/rest/classes/ezpkernelrest.php]: Why not having it?

Comment (bdunogier) [inline kernel/private/rest/classes/ezpkernelrest.php]: I don't know. I'm pretty sure it is a copy/paste from another kernel :p

File: index_rest.php (hunk 2 above)
Comment (lolautruche) [inline index_rest.php]: Then remove it :-)

File: kernel/private/rest/classes/ezpkernelrest.php (hunk 1 above)
Comment (lolautruche) [inline kernel/private/rest/classes/ezpkernelrest.php]: yep

**Comment by bdunogier** [comment]
I'm gonna do a cleanup pass on the existing comments this morning.

File: kernel/private/rest/classes/ezpkernelrest.php (hunk 6 above)
Comment (lolautruche) [inline kernel/private/rest/classes/ezpkernelrest.php]: You should specify a default value for `$responseWriterClass` as it comes after an argument with default value.

Comment (lolautruche) [inline kernel/private/rest/classes/ezpkernelrest.php]: Besides, no argument is specified at all in `index_rest.php`.

File: kernel/private/rest/classes/ezpkernelrest.php [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+        self::$response = $response;
+    }
+
+    /**
+     * Initializes the session. If running, through Symfony the session
+     * parameters from Symfony override the session parameter from eZ Publish.
+     */
+    protected function sessionInit()
+    {
+    }
```
Comment (lolautruche) [inline kernel/private/rest/classes/ezpkernelrest.php]: I guess you don't need this method?

File: kernel/private/rest/classes/ezpkernelrest.php [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+     */
+    public function reInitialize()
+    {
+        $this->isInitialized = false;
+    }
+
+    /**
+     * Runs the shutdown process
+     */
+    protected function shutdown( $reInitialize = true )
```
Comment (lolautruche) [inline kernel/private/rest/classes/ezpkernelrest.php]: Doesn't seem to be used. Missing call in `run()`?

File: kernel/private/rest/classes/ezpkernelrest.php [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+     * @param string $errorMessage
+     * @param int $errorCode
+     * @return ezpKernelResult
+     *
+     * @throws RuntimeException
+     * @throws ezpAccessDenied
+     *
+     * @see setUseExceptions()
+     */
+    protected function exitWithInternalError( $errorMessage, $errorCode = 500 )
```
Comment (lolautruche) [inline kernel/private/rest/classes/ezpkernelrest.php]: Doesn't seem to be used

File: kernel/private/rest/classes/ezpkernelrest.php (hunk 8 above)
Comment (lolautruche) [inline kernel/private/rest/classes/ezpkernelrest.php]: For us: nothing. But can be useful using custom code... The same goes for `ezpKernelWeb` :wink:

**Comment by lolautruche** [comment]
Apart from needed cleanup and my small comments: +1. Good job !

Did you test on a pure legacy stack?

**Comment by bdunogier** [comment]
> Did you test on a pure legacy stack?

I think I have, yes.

File: kernel/private/rest/classes/ezpkernelrest.php (hunk 6 above)
Comment (bdunogier) [inline kernel/private/rest/classes/ezpkernelrest.php]: I'm wondering why it works, actually... :-)

File: kernel/private/rest/classes/ezpkernelrest.php (hunk 10 above)
Comment (bdunogier) [inline kernel/private/rest/classes/ezpkernelrest.php]: Fixed.

**Comment by bdunogier** [comment]
All remarks fixed, and rebased.

**Comment by yannickroger** [comment]
+1

**Pull request merged**